File: mafbase_stream/android/src/main/kotlin/com/example/mafbase_stream/overlay/OverlayCatalog.kt
package com.example.mafbase_stream.overlay

import android.content.Context
import android.view.View
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.example.mafbase_stream.overlay.plashki.PlashkiMafbaseOverlay

/**
 * Внутренний каталог overlay-вёрстки, известной плагину. Каждый overlay — это
 * `@Composable`, принимающий [OverlayParams], а [OverlayComposeContainer]
 * оборачивает его в `ComposeView` со всем нужным окружением.
 *
 * Чтобы добавить новый overlay — пиши `@Composable` функцию, принимающую
 * [OverlayParams], и добавляй строчку в [overlays] под уникальным `viewType`.
 * Тот же `viewType` приложение указывает из Dart через `MafbaseOverlay`.
 *
 * Помимо overlay-вёрстки каталог рисует [BrandImageLayer] под ней — постоянная
 * брендированная картинка, заданная через [OverlayParams.brandImageUrl]. Она
 * работает независимо от того, выбран overlay или нет.
 */
internal object OverlayCatalog {
    private val overlays: Map<String, @Composable (OverlayParams) -> Unit> = mapOf(
        "plashki-mafbase" to { params -> PlashkiMafbaseOverlay(params) },
    )

    fun has(viewType: String): Boolean = viewType in overlays

    /**
     * Создаёт wrapper-view с brand-слоем и (опционально) overlay-вёрсткой.
     * Возвращает null, если ни overlay, ни brand не заданы — тогда смысла
     * подключать overlay-renderer нет.
     */
    fun create(
        viewType: String?,
        context: Context,
        invalidator: OverlayInvalidator,
        params: OverlayParams = OverlayParams(),
    ): View? {
        val overlayContent: (@Composable (OverlayParams) -> Unit)? = viewType?.let { overlays[it] }
        if (overlayContent == null && params.brandImageUrl.isNullOrBlank()) return null
        val wrappedContent: @Composable (OverlayParams) -> Unit = { p ->
            Box(Modifier.fillMaxSize()) {
                BrandImageLayer(p.brandImageUrl)
                overlayContent?.invoke(p)
            }
        }
        return OverlayComposeContainer(context, invalidator, params, wrappedContent)
    }
}

File: mafbase_stream/ios/Classes/Overlay/OverlayCatalog.swift
import UIKit

/// Внутренний каталог overlay-вёрстки, известной плагину. Каждый overlay — это
/// фабрика `(OverlayParams, OverlayInvalidator) -> UIView`, которую плагин знает
/// под уникальным `viewType`.
///
/// Чтобы добавить новый overlay — реализуй UIView, принимающую `OverlayParams` +
/// `OverlayInvalidator`, и зарегистрируй её в `overlays`. Тот же `viewType`
/// приложение указывает из Dart через `MafbaseOverlay`.
///
/// Помимо overlay'я каталог рисует [BrandOverlayContainerView] под ним —
/// постоянная брендированная картинка, заданная через `params.brandImageUrl`.
/// Она работает независимо от того, выбран overlay или нет.
enum OverlayCatalog {

    typealias Factory = (OverlayParams, OverlayInvalidator) -> UIView

    private static let overlays: [String: Factory] = [
        "plashki-mafbase": { params, invalidator in
            PlashkiMafbaseOverlay(params: params, invalidator: invalidator)
        },
    ]

    static func has(viewType: String) -> Bool {
        return overlays[viewType] != nil
    }

    /// Создаёт wrapper с brand-слоем и (опционально) overlay-вёрсткой.
    /// Возвращает nil, если ни overlay, ни brand-image не заданы — overlay-renderer
    /// в таком случае подключать не нужно.
    static func create(
        viewType: String?,
        params: OverlayParams,
        invalidator: OverlayInvalidator
    ) -> UIView? {
        let overlayView: UIView? = viewType.flatMap { overlays[$0] }.map { $0(params, invalidator) }
        let hasBrand = (params.brandImageUrl?.isEmpty == false)
        if overlayView == nil && !hasBrand {
            return nil
        }
        return BrandOverlayContainerView(
            brandImageUrl: params.brandImageUrl,
            overlayView: overlayView,
            invalidator: invalidator
        )
    }
}
